salt/modules/smartos_mdata.py

```diff
@@ -0,0 +1,184 @@
+    looks to see if mdata-list is present on the system
+    '''
+    return salt.utils.which('mdata-list')
+
+
+@decorators.memoize
+def _check_mdata_get():
+    '''
+    looks to see if mdata-get is present on the system
+    '''
+    return salt.utils.which('mdata-get')
+
+
+@decorators.memoize
+def _check_mdata_put():
+    '''
+    looks to see if mdata-put is present on the system
+    '''
+    return salt.utils.which('mdata-put')
+
+
+@decorators.memoize
+def _check_mdata_delete():
+    '''
+    looks to see if mdata-delete is present on the system
+    '''
+    return salt.utils.which('mdata-delete')
+
+
+def __virtual__():
+    '''
+    Provides mdata only on SmartOS
+    '''
+    if salt.utils.is_smartos_zone() and _check_mdata_list():
+        return __virtualname__
+    return (
+        False,
+        '{0} module can only be loaded on SmartOS zones'.format(
+            __virtualname__
+        )
+    )
+
+
+def list_():
+    '''
+    List available metadata
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' mdata.list
+    '''
+    mdata = _check_mdata_list()
+    if mdata:
+        cmd = '{0}'.format(mdata)
+        return __salt__['cmd.run'](cmd).splitlines()
+    return {}
+
+
+def get_(*keyname):
+    '''
+    Get metadata
+
+    keyname : string
+        name of key
+
+    .. note::
+
+        If no keynames are specified, we get all properties
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' mdata.get salt:role
+        salt '*' mdata.get user-script salt:role
+    '''
+    mdata = _check_mdata_get()
+    valid_keynames = list_()
+    ret = {}
+
+    if len(keyname) == 0:
+        keyname = valid_keynames
+
+    for k in keyname:
+        if mdata and k in valid_keynames:
+            cmd = '{0} {1}'.format(mdata, k)
+            ret[k] = __salt__['cmd.run'](cmd)
+        else:
+            ret[k] = ''
+
+    return ret
+
+
+def put_(keyname, val):
+    '''
+    Put metadata
+
+    prop : string
+        name of property
+    val : string
+        value to set
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' mdata.list
+    '''
+    mdata = _check_mdata_put()
+    ret = {}
+
+    if mdata:
+        cmd = 'echo {2} | {0} {1}'.format(mdata, keyname, val)
+        ret = __salt__['cmd.run_all'](cmd, python_shell=True)
+
+    return ret['retcode'] == 0
+
+
+def delete_(*keyname):
+    '''
+    Delete metadata
+
+    prop : string
+        name of property
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' mdata.get salt:role
+        salt '*' mdata.get user-script salt:role
+    '''
+    mdata = _check_mdata_delete()
+    valid_keynames = list_()
+    ret = {}
+
+    for k in keyname:
+        if mdata and k in valid_keynames:
+            cmd = '{0} {1}'.format(mdata, k)
+            ret[k] = __salt__['cmd.run_all'](cmd)['retcode'] == 0
+        else:
+            ret[k] = True
+
+    return ret
+
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
```

Comment:
Actually makes it a bit easier for vim users as it forced vim to use the tab settings specified in the file and not what the user has setup.